### PR TITLE
Fixed #28752 -- Made django.setup() idempotent.

### PR DESCRIPTION
**Pull request by pakal**
Stripped-down version of ticket #30536, without tweakable django setup.

### CODE REVIEW

File: django/__init__.py [hunk 1]
```diff
@@ -5,12 +5,26 @@
...
 def setup(set_prefix=True):
     """
     Configure the settings (this happens as a side effect of accessing the
     first setting), configure logging and populate the app registry.
     Set the thread-local urlresolvers script prefix if `set_prefix` is True.
+
+    This function is idempotent (calling it several times changes nothing), but
+    not reentrant (recursively calling it would have an unexpected result).
+
+    Returns True iff actual operations occurred.
```
Comment (adamchainz) [inline django/__init__.py]: I think the current wording is a bit hard to follow:
* "iff" is a rare, more academic word, that won't translate well
* "actual operations occurred" is a bit vague
* The case where "actual operations didn't occur" is not clearly specified - will `None` be returned.

How about "Returns a boolean indicating whether setup occurred for the first time" ?

File: django/__init__.py [hunk 2]
```diff
@@ -22,3 +36,7 @@ def setup(set_prefix=True):
             '/' if settings.FORCE_SCRIPT_NAME is None else settings.FORCE_SCRIPT_NAME
         )
     apps.populate(settings.INSTALLED_APPS)
+
+    # Only now that setup was successful, we prevent later recalls
```
Comment (adamchainz) [inline django/__init__.py]: Not sure this comment really helps since we already know what `is_ready` is for

File: docs/releases/3.0.txt [hunk 3]
```diff
@@ -131,6 +131,11 @@ Minor features
   The default value is :func:`~django.utils.translation.get_language()` instead
   of :setting:`LANGUAGE_CODE`.
 
+App-loading changes
+~~~~~~~~~~~~~~~~~~~~
+
+* ``django.setup()`` is now idempotent, and returns True if and only if actual operations were performed.
```
Comment (adamchainz) [inline docs/releases/3.0.txt]: As above, I'd suggest

```
``django.setup()`` is now idempotent, and returns a boolean indicating whether setup occurred on this call.
```

File: django/__init__.py (hunk 1 above)
Comment (pakal) [reply django/__init__.py]: Ok changed

File: django/__init__.py (hunk 2 above)
Comment (pakal) [reply django/__init__.py]: Ok changed

File: docs/releases/3.0.txt (hunk 3 above)
Comment (pakal) [reply docs/releases/3.0.txt]: Ok changed

**Comment by pakal** [comment]
As suggested in https://github.com/django/django/pull/11435, I protected the setup with a threading lock, it avoids nasty race conditions when people try to lazy-setup it concurrently in a server, and makes sure "is_ready" doesn't get abused.

File: docs/topics/settings.txt [hunk 4]
```diff
@@ -291,16 +291,20 @@ Note that calling ``django.setup()`` is only necessary if your code is truly
...
-
-    Therefore, avoid putting reusable application logic in standalone scripts
-    so that you have to import from the script elsewhere in your application.
-    If you can't avoid that, put the call to ``django.setup()`` inside an
-    ``if`` block::
-
-        if __name__ == '__main__':
-            import django
-            django.setup()
+Note that ``django.setup()`` is idempotent - all calls after the first one will
```
Comment (pakal) [inline docs/topics/settings.txt]: Maybe I should update this string too, regarding thread-safety and return value.

Comment (pakal) [reply docs/topics/settings.txt]: Fixed

**Comment by pakal** [comment]
OK I had used different locks during django.setup(), so thread-safety wasn't OK... I just fixed that and reinforced concurrency tests.

**Comment by pakal** [comment]
Yes in some cases it might already be idempotent, but lock-less idempotence is like lock-less threadsafety,  the tiniest change in the code might break it silently. ^^
I guess a top-level mutex is a cheap and robust way to protect all this critical phase of setup (eg. if someone tweaks logging conf after setup(), another call to setup() could revert his changes).   

I'll have a look asap at deduplicating locks, and raising in case of recursive calls.

**Comment by aaugustin** [comment]
Agreed, let's avoid relying on weak implementation details.

What I'm really asking here is — why are we making this change?

_"This doesn't change anything in practice, but I find it more future-proof to move the lock from Apps.populate() to django.setup()"_ is one possible answer.

This isn't where the discussion started, though, so I'd like to make sure we aren't working on a patch that doesn't actually address the original issue.

**Comment by pakal** [comment]
Now that I remember, originally the idempotence idea arised for me from some event where I ended up having some loggers erased by those of LOGGING, due to accidental multiple setup() calls. 

I've fixed the reentracy of django.setup() in my PR.  :)

Concerning that of populate(), I think that it could be rather dangerous and unneeded to remove it, since it's documented (so it would be a backwards incompatible change), and we never know in what buggy occasion populate() calls could also recursively, or who might have needed to call populate() aside Django's setup. 
Having threadsafe configuration utilities is always a good thing imho, considered the cheapness of mutexes (we could even put a threadsafety decorator on individual configuration functions, even "configure_logging()", but it's another debate), and the diversity of the runners/apps ecosystem (which might sometimes attempt weird stuffs at startup).

Would the PR be OK this way? Or is it a problem to let these multiple protections?

**Comment by pakal** [comment]
Damn, everytime I push, a different CI check, for a different reason unrelated to the django tests (failed checkout, xmlrunner missing...) ; is there a way to force a restart on a single check ?

**Comment by pakal** [comment]
OK all checks passing now

**Comment by aaugustin** [comment]
> Having threadsafe configuration utilities is always a good thing imho, considered the cheapness of mutexes

Mutexes create risks of deadlocks, which is expensive from a user experience and project maintenance perspective.

> (we could even put a threadsafety decorator on individual configuration functions, even "configure_logging()", but it's another debate)

It isn't another debate. It's the actual bug here and the correct fix!

> and the diversity of the runners/apps ecosystem (which might sometimes attempt weird stuffs at startup).

This is exactly why your proposal is wrong.

Since we're assuming that weird stuff may happen:

- a thread could call `django.setup()` which will call `apps.populate()`
- another thread could call `apps.populate()`; while it executes, it could hit a call to `django.setup()`

There you have a deadlock scenario.

Scattershot changes that have the side effect of doing something you want aren't the correct way to fix bugs in this area. Please think about the unintended consequences of the changes you're proposing.

**Comment by pakal** [comment]
Indeed I had overlooked that scenario of jumping upwards from populate() to setup(), my bad.

There are two risks I see here: breaking the assumptions of people relying on populate()'s current threadsafety/idempotence (if its lock is removed), and ending up with new deadlocks (if its lock is left as is).

In your googlegroups answer (https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/django-developers/tePzJ3Azwnk/f4Im9wfiCAAJ) you evocate fixing only configure_logging() idempotence, but it seems to me that we have the same problem: if we use locks, we might have deadlocks, and if we don't, idempotence will be easily ruined by race conditions, in such heavily threaded environment.

How about an alternate implementation, where all initialization utilities (setup(), and nested utilities like configure_logging and populate()) protect themselves with a shared (and recursive) "setup lock" ? Thus, we avoid deadlocks, and "weird" code fiddling with separate initialization functions will keep working, and future evolutions of setup() are protected against race-condition/idempotence regressions. It'll be worth auditing existing locks of Django in that PR, to ensure deadlock risks are almost non-existent. What about this way of proceeding?

I acknowledge that this PR is a bit fuzzy - back when I had troubles with setup() I had no time at all to contribute tickets and investigate global solutions, and now I don't have a precise bug case blocking me; but we see why applying a logging configuration several times might reset/break logger configuration tweaks issued from other parts of the code (eg. especially non-django modules). 
This PR is almost more an enhancement than a bugfix, since it removes from external packages the burden of caring against multiple setup() calls, and prevents future (silent) regressions when setup() might be extended with new steps. For me having django.setup() not globally protected is a bit like having the __init__() method of a class called multiple times - it might mostly work *for now*, but it's still a silent threat hovering on all kinds of future evolutions.

**Comment by aaugustin** [comment]
> In your googlegroups answer (https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/django-developers/tePzJ3Azwnk/f4Im9wfiCAAJ) you evocate fixing only configure_logging() idempotence, but it seems to me that we have the same problem: if we use locks, we might have deadlocks, and if we don't, idempotence will be easily ruined by race conditions, in such heavily threaded environment.

No, we don't. In my proposal, the order is:

1. Take "configure logging" lock
2. Release "configure logging" lock
3. Take "populate apps" lock
4. Release "populate apps" lock

This isn't vulnerable to deadlocks because locks aren't nested.

Consider the difference with:

1. Take "setup Django" lock
2. Take "populate apps" lock
3. Release "populate apps" lock
4. Release "setup Django" lock

**Comment by pakal** [comment]
I see, I was still in the perspective "very weird stuffs happening", in which if a populate() can call django.setup(), then configure_logging() and populate() might as well call each other (although probabilities of this happening are indeed becoming nano-scale).

In the end, how about my proposal of protecting all these utilities (django.setup AND each specific initialization function) with a common "initialization lock", so that both the common scenario and direct fiddlings get protected against concurrent access + replay + deadlocks?

Or is there a reason to separate threading locks, and leave django.setup() replayable?

**Comment by aaugustin** [comment]
I suspect it will be hard to keep the current behavior of raising an exception on re-entrant calls to `Apps.populate()` if you're already taking the same lock earlier in `django.setup()`, but I didn't try it, so I might be wrong.

----

Stepping back for a minute, I think one of the reasons we're talking past each other is that we haven't agreed on the goal we're pursuing here.

Django documents public APIs and guarantees some behaviors as long as you stay within documented use of these APIs. This is why I've been asking for concrete use cases. I would like to determine what behaviors we want to guarantee and then to make the code match these requirements. I'm finding it hard to have this conversation when I don't know what the success criterion is. How do we know that this issue is fixed? How do we test it?

You're trying to ensure that Django works even if "weird stuff happens", including arbitrary monkey-patching. The line between "weird but acceptable" and "so weird it can't be supported" is fuzzy. Whatever pytest-django does mostly works currently. Far-fetched example:`import django; django.setup = None` goes too far. If we can't document supported usage of the APIs, then we can't reason about whether the implementation of Django provides the intended behavior, because we don't know enough about the context.

**Comment by pakal** [comment]
I think that as long as populate() keeps its local "self.loading" boolean flag, it will be able to detect an abnormal reentrant call, and raise as it currently does.

My intention is not to prevent every "weird stuff", just to ensure initialization routines have these features:

- being thread safe; most of the framework code doesn't need any special measure for that, since it only modifies local or threadlocal variables, but initialization functions have to fiddle with global process state (logging, prefix, app registry...) so locks seem unavoidable if we want a simple and robust way (safe lock-less concurrent access is an art I'd rather not touch).

- being only run once in the whole process life, since multiple runs of initializers reset some configuration tweaks that the user could have made inbetween, and can breaks stuff in subtle ways.

These reasons, if I understood correctly, are already those that lead to populate() being protected with locks ; I'd just like to extend this protection to the whole django.setup() workflow, primarily so that configure_logging() can't be executed several times anymore, secondly so that no subtle regression occurs in the future when (almost unavoidably) new steps will be added to the django setup phase.

Separately locking each big step (configure_logging(), populate()...) would already do the job; I'm just wondering if it wouldn't be aven better to use a common lock so that *even* if weird interactions occur, no lock nesting occurs, it's only the same recursive lock which gets double-taken by the only authorized thread. But this might be considered over-complication for a nanoscopic risk of lock cross-nesting.

**Comment by pakal** [comment]
Any update on this PR? Is anything blocking it?

It seems to me that having the whole initialization of a web framework - which runs in often heavily threaded environments - threadsafe, is a rather consensual feature, isn't it?

**Comment by pope1ni** [comment]
You need to set _"Has patch"_ and unset _"Patch needs improvement"_ on the [ticket](https://code.djangoproject.com/ticket/28752) to add it to the review queue.

**Comment by aaugustin** [comment]
> Is anything blocking it?

Yes, the fact that you still haven't provided practical use cases.

**Comment by carltongibson** [comment]
Yes, I need to review but AFAICS there’s not consensus about how or whether to proceed. (Which is for the mailing list discussion rather than here.)

**Comment by pakal** [comment]
There is a discusion here for both current improvements attempts of django setup (idempotence and tweakability), please voice your design concerns if any : https://groups.google.com/forum/#!topic/django-developers/tePzJ3Azwnk

Thanks @pope1nin I updated the ticket ( https://code.djangoproject.com/ticket/28752 )

@aaugustin - what do you mean by "practical use case"? This PR is about preventing misc bugs (see tickets) and making the Django setup robust and future-proof, not adding capabilities. The rationale is the same as that which lead to protecting apps.populate(), I just propose to extend this basic behaviour to the whole Django setup. If from the start django.setup() had been idempotent and exposed an "is_setup" indicator, no one would have blinked or questioned it, so why is it so controversial nowadays?

**Comment by aaugustin** [comment]
> misc bugs (see tickets)

The closest thing to a bug I saw in a ticket was misuse of a private API by pytest-django, which doesn't count as a bug.

**Comment by aaugustin** [comment]
I don't think I'm helping here so I'm going to unsubscribe from this PR.

I still think there's exactly one bug worth fixing in this area: `configure_logging()` is vulnerable to concurrent calls. It should be made idempotent and protected with a lock.

**Comment by pakal** [comment]
Yes we could fix *this* current issue (silent multiple overwriting of logging conf) by protecting configure_logging(). I just don't get what we gain by *not* protecting the whole setup() at once too... any thoughts anyone ?

**Comment by carltongibson** [comment]
Closing as per ticket.

**Pull request closed without merging**